modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeJavaMultiVMPoolUtilCheck.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.source.formatter.check;
+
+import com.liferay.petra.string.CharPool;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.source.formatter.check.util.JavaSourceUtil;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author Nícolas Moura
+ */
+public class UpgradeJavaMultiVMPoolUtilCheck extends BaseFileCheck {
+
+	@Override
+	protected String doProcess(
+			String fileName, String absolutePath, String content)
+		throws Exception {
+
+		if (!fileName.endsWith(".java")) {
+			return content;
+		}
+
+		if (content.contains(_MULTI_VM_POOL_UTIL_IMPORT)) {
+			content = StringUtil.replace(
+				content, _MULTI_VM_POOL_UTIL_IMPORT,
+				"import com.liferay.portal.kernel.cache.MultiVMPool;");
+			content = _replaceGetPortalCache(content, fileName);
+		}
+
+		return content;
+	}
+
+	private String _replaceGetPortalCache(String content, String fileName)
+		throws Exception {
+
+		content = JavaSourceUtil.addImport(
+			content,
+			"import org.osgi.service.component.annotations.Reference;");
+
+		Matcher portalCacheMatcher = _getPortalCachePattern.matcher(content);
+
+		while (portalCacheMatcher.find()) {
+			String newDeclaration = StringUtil.replace(
+				portalCacheMatcher.group(0), "MultiVMPoolUtil.getPortalCache(",
+				"(PortalCache" + portalCacheMatcher.group(1) +
+					") _multiVMPool.getPortalCache(");
+
+			content = StringUtil.replace(
+				content, portalCacheMatcher.group(0), newDeclaration);
+		}
+
+		content = StringUtil.replace(
+			content, "MultiVMPoolUtil.getPortalCache(",
+			_WARNING_CASE + " _multiVMPool.getPortalCache(");
+		content = StringUtil.replaceLast(
+			content, CharPool.CLOSE_CURLY_BRACE,
+			"\n\t@Reference\n\tprivate MultiVMPool _multiVMPool;\n}");
+
+		addMessage(
+			fileName, "Replace 'TO_BE_REPLACED' with the correct type",
+			getLineNumber(content, content.indexOf(_WARNING_CASE)));
+
+		return content;
+	}
+
+	private static final String _MULTI_VM_POOL_UTIL_IMPORT =
+		"import com.liferay.portal.kernel.cache.MultiVMPoolUtil;";
+
+	private static final String _WARNING_CASE =
+		"(PortalCache<TO_BE_REPLACED, TO_BE_REPLACED>)";
```

Comment:
Can we rename to something like `_DEFAULT_TYPE`?

Also, I think we discussed this earlier, but is there an issue if we use `PortalCache<? extends Serializable, ? extends Serializable>` instead?

Reply:
Yeah, this suggestion can lead to some compilation problems on client code, needs to be the exact type. And this was made this way to make the warning explicit, yeah. And of course, we can rename the variable.